src/proficiency_ui.cpp

```diff
@@ -0,0 +1,375 @@
+    right_print( w_header, 0, 1, c_white, hint_txt );
+    wnoutrefresh( w_header );
+}
+
+void _prof_window::draw_borders()
+{
+    const int w = catacurses::getmaxx( w_border );
+    const int h = catacurses::getmaxy( w_border );
+    draw_border( w_border, c_white, _( "Proficiencies" ), c_yellow );
+    // horizontal header separator
+    for( int i = 1; i < w - 1; i++ ) {
+        mvwputch( w_border, point( i, 4 ), c_white, LINE_OXOX );
+    }
+    mvwputch( w_border, point( 0, 4 ), c_white, LINE_XXXO );
+    mvwputch( w_border, point( w - 1, 4 ), c_white, LINE_XOXX );
+    // vertical column separator
+    for( int i = 5; i < h - 1; i++ ) {
+        mvwputch( w_border, point( column_width + 1, i ), c_white, LINE_XOXO );
+    }
+    mvwputch( w_border, point( column_width + 1, 4 ), c_white, LINE_OXXX );
+    mvwputch( w_border, point( column_width + 1, h - 1 ), c_white, LINE_XXOX );
+
+    scrollbar()
+    .border_color( c_white )
+    .offset_x( 0 )
+    .offset_y( 5 )
+    .content_size( get_current_set().size() )
+    .viewport_pos( top_prof )
+    .viewport_size( max_rows )
+    .apply( w_border );
+
+    wnoutrefresh( w_border );
+}
+
+void _prof_window::init_ui_windows()
+{
+    const int h = std::min( 28, TERMY );
+    const int w = std::min( 120, TERMX );
+    const point origin( TERMX / 2 - w / 2, TERMY / 2 - h / 2 );
+    const int center = w / 2;
+    max_rows = h - 6;
+    column_width = center - 2;
+
+    w_border  = catacurses::newwin( h, w, origin );
+    //NOLINTNEXTLINE(cata-use-named-point-constants)
+    w_header  = catacurses::newwin( 3, w - 2, origin + point( 1, 1 ) );
+    w_profs   = catacurses::newwin( max_rows, column_width, origin + point( 1, 5 ) );
+    w_details = catacurses::newwin( max_rows, column_width, origin + point( center + 1, 5 ) );
+}
+
+shared_ptr_fast<ui_adaptor> _prof_window::create_or_get_ui_adaptor()
+{
+    shared_ptr_fast<ui_adaptor> current_ui = ui.lock();
+    if( !current_ui ) {
+        ui = current_ui = make_shared_fast<ui_adaptor>();
+        current_ui->on_screen_resize( [this]( ui_adaptor & cui ) {
+            init_ui_windows();
+            cui.position_from_window( catacurses::stdscr );
+        } );
+        current_ui->mark_resize();
+        current_ui->on_redraw( [this]( const ui_adaptor & ) {
+            draw_borders();
+            draw_header();
+            draw_profs();
+            draw_details();
+        } );
+    }
+    return current_ui;
+}
+
+void _prof_window::run()
+{
+    if( !u ) {
+        return;
+    }
+
+    ctxt = input_context( "PROFICIENCY_WINDOW" );
+    ctxt.register_directions();
+    ctxt.register_action( "HELP_KEYBINDINGS" );
+    ctxt.register_action( "PREV_CATEGORY" );
+    ctxt.register_action( "NEXT_CATEGORY" );
+    ctxt.register_action( "FILTER" );
+    ctxt.register_action( "QUIT" );
+
+    populate_categories();
+    shared_ptr_fast<ui_adaptor> current_ui = create_or_get_ui_adaptor();
+    init_ui_windows();
```

Comment:
There's no need to call it here because `ui_manager::redraw` will call it.

Reply:
```suggestion
```